File: openags/research/messaging/feishu.py
"""Feishu (Lark) notification channel via Bot Webhook."""

from __future__ import annotations

import logging
import time

import httpx

from openags.models import FeishuConfig

logger = logging.getLogger(__name__)

_LEVEL_EMOJI = {
    "info": "ℹ️",
    "warning": "⚠️",
    "error": "❌",
    "success": "✅",
}


class FeishuChannel:
    """Send notifications to Feishu/Lark via Bot Webhook or API.

    Supports two modes:
      1. Webhook URL (simple, no OAuth needed)
      2. App API with app_id/app_secret (for richer message types)

    For webhook mode, set chat_id to the webhook URL.
    For API mode, provide app_id and app_secret.
    """

    def __init__(self, config: FeishuConfig) -> None:
        self._app_id = config.app_id
        self._app_secret = config.app_secret.get_secret_value()
        self._chat_id = config.chat_id
        self._tenant_token: str | None = None
        self._token_expires: float = 0

    @property
    def channel_id(self) -> str:
        return "feishu"

    async def send_text(self, text: str) -> bool:
        return await self._send_webhook({"msg_type": "text", "content": {"text": text}})

    async def send_markdown(self, markdown: str) -> bool:
        # Feishu uses "interactive" card for rich formatting
        card = {
            "msg_type": "interactive",
            "card": {
                "elements": [
                    {
                        "tag": "markdown",
                        "content": markdown,
                    }
                ],
            },
        }
        return await self._send_webhook(card)

    async def send_notification(
        self,
        title: str,
        body: str,
        level: str = "info",
    ) -> bool:
        emoji = _LEVEL_EMOJI.get(level, "ℹ️")
        card = {
            "msg_type": "interactive",
            "card": {
                "header": {
                    "title": {"tag": "plain_text", "content": f"{emoji} {title}"},
                    "template": "blue" if level == "info" else "red" if level == "error" else "orange",
                },
                "elements": [
                    {"tag": "markdown", "content": body},
                ],
            },
        }
        return await self._send_webhook(card)

    async def health_check(self) -> bool:
        """Check connectivity by getting a tenant access token."""
        try:
            token = await self._get_tenant_token()
            return token is not None
        except Exception as e:
            logger.warning("Feishu health check failed: %s", e)
            return False

    async def _send_webhook(self, payload: dict) -> bool:
        """Send via webhook URL (chat_id as webhook URL)."""
        url = self._chat_id
        if not url.startswith("http"):
            # If chat_id is not a URL, use the API method
            return await self._send_api(payload)

        try:
            async with httpx.AsyncClient(timeout=15) as client:
                r = await client.post(url, json=payload)
                data = r.json()
                if data.get("code") == 0 or data.get("StatusCode") == 0:
                    return True
                logger.warning("Feishu webhook failed: %s", data)
                return False
        except Exception as e:
            logger.error("Feishu send error: %s", e)
            return False

    async def _send_api(self, payload: dict) -> bool:
        """Send via Feishu API with tenant token."""
        token = await self._get_tenant_token()
        if not token:
            return False

        url = "https://open.feishu.cn/open-apis/im/v1/messages"
        params = {"receive_id_type": "chat_id"}
        headers = {"Authorization": f"Bearer {token}"}

        body = {
            "receive_id": self._chat_id,
            **payload,
        }

        try:
            async with httpx.AsyncClient(timeout=15) as client:
                r = await client.post(url, params=params, headers=headers, json=body)
                data = r.json()
                if data.get("code") == 0:
                    return True
                logger.warning("Feishu API send failed: %s", data)
                return False
        except Exception as e:
            logger.error("Feishu API error: %s", e)
            return False

    async def _get_tenant_token(self) -> str | None:
        """Get or refresh the Feishu tenant access token."""
        if self._tenant_token and time.time() < self._token_expires:
            return self._tenant_token

        url = "https://open.feishu.cn/open-apis/auth/v3/tenant_access_token/internal"
        payload = {
            "app_id": self._app_id,
            "app_secret": self._app_secret,
        }

        try:
            async with httpx.AsyncClient(timeout=10) as client:
                r = await client.post(url, json=payload)
                data = r.json()
                if data.get("code") == 0:
                    self._tenant_token = data["tenant_access_token"]
                    self._token_expires = time.time() + data.get("expire", 7200) - 300
                    return self._tenant_token
        except Exception as e:
            logger.error("Feishu token error: %s", e)

        return None

File: packages/app/src/messaging/feishu.ts
/**
 * Feishu (Lark) Bot Integration
 *
 * Send messages via Feishu webhook or Bot API.
 */

export interface FeishuConfig {
  /** Webhook URL for simple messaging */
  webhookUrl?: string
  /** App ID for full API access */
  appId?: string
  /** App Secret for full API access */
  appSecret?: string
  /** Default chat ID for notifications */
  chatId?: string
}

export interface FeishuTextMessage {
  msg_type: 'text'
  content: {
    text: string
  }
}

export interface FeishuPostMessage {
  msg_type: 'post'
  content: {
    post: {
      zh_cn?: FeishuPostContent
      en_us?: FeishuPostContent
    }
  }
}

export interface FeishuPostContent {
  title: string
  content: Array<Array<FeishuPostElement>>
}

export type FeishuPostElement =
  | { tag: 'text'; text: string }
  | { tag: 'a'; text: string; href: string }
  | { tag: 'at'; user_id: string }
  | { tag: 'img'; image_key: string }

export type FeishuCardColor = 'blue' | 'wathet' | 'turquoise' | 'green' | 'yellow' | 'orange' | 'red' | 'carmine' | 'violet' | 'purple' | 'indigo' | 'grey'

export interface FeishuCardMessage {
  msg_type: 'interactive'
  card: {
    header?: {
      title: {
        tag: 'plain_text'
        content: string
      }
      template?: FeishuCardColor
    }
    elements: FeishuCardElement[]
  }
}

export type FeishuCardElement =
  | { tag: 'div'; text: { tag: 'plain_text' | 'lark_md'; content: string } }
  | { tag: 'hr' }
  | { tag: 'note'; elements: Array<{ tag: 'plain_text' | 'lark_md'; content: string }> }

export type FeishuMessage = FeishuTextMessage | FeishuPostMessage | FeishuCardMessage

const FEISHU_API = 'https://open.feishu.cn/open-apis'

export class FeishuBot {
  private config: FeishuConfig
  private accessToken: string | null = null
  private tokenExpiry: number = 0

  constructor(config: FeishuConfig) {
    this.config = config
  }

  /**
   * Send a message via webhook.
   */
  async sendWebhook(message: FeishuMessage): Promise<boolean> {
    if (!this.config.webhookUrl) {
      throw new Error('webhookUrl not configured')
    }

    const response = await fetch(this.config.webhookUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(message),
    })

    const result = await response.json() as { code: number }
    return result.code === 0
  }

  /**
   * Send a simple text notification.
   */
  async notify(text: string): Promise<boolean> {
    if (this.config.webhookUrl) {
      return this.sendWebhook({
        msg_type: 'text',
        content: { text },
      })
    }

    // Fall back to bot API
    if (this.config.appId && this.config.appSecret && this.config.chatId) {
      return this.sendMessage(this.config.chatId, {
        msg_type: 'text',
        content: { text },
      })
    }

    throw new Error('No webhook URL or app credentials configured')
  }

  /**
   * Send a rich card notification.
   */
  async notifyCard(title: string, content: string, color?: FeishuCardColor): Promise<boolean> {
    const card: FeishuCardMessage = {
      msg_type: 'interactive',
      card: {
        header: {
          title: { tag: 'plain_text', content: title },
          template: color || 'blue',
        },
        elements: [
          { tag: 'div', text: { tag: 'lark_md', content } },
        ],
      },
    }

    if (this.config.webhookUrl) {
      return this.sendWebhook(card)
    }

    if (this.config.appId && this.config.appSecret && this.config.chatId) {
      return this.sendMessage(this.config.chatId, card)
    }

    throw new Error('No webhook URL or app credentials configured')
  }

  /**
   * Get tenant access token for API calls.
   */
  private async getAccessToken(): Promise<string> {
    if (this.accessToken && Date.now() < this.tokenExpiry) {
      return this.accessToken
    }

    if (!this.config.appId || !this.config.appSecret) {
      throw new Error('appId and appSecret required for API calls')
    }

    const url = `${FEISHU_API}/auth/v3/tenant_access_token/internal`

    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        app_id: this.config.appId,
        app_secret: this.config.appSecret,
      }),
    })

    const result = await response.json() as {
      code: number
      tenant_access_token: string
      expire: number
    }

    if (result.code !== 0) {
      throw new Error('Failed to get access token')
    }

    this.accessToken = result.tenant_access_token
    // Expire 5 minutes early to be safe
    this.tokenExpiry = Date.now() + (result.expire - 300) * 1000

    return this.accessToken
  }

  /**
   * Send a message via bot API.
   */
  async sendMessage(chatId: string, message: FeishuMessage): Promise<boolean> {
    const token = await this.getAccessToken()
    const url = `${FEISHU_API}/im/v1/messages?receive_id_type=chat_id`

    const body = {
      receive_id: chatId,
      msg_type: message.msg_type,
      content: JSON.stringify(message.msg_type === 'interactive' ? message.card : message.content),
    }

    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${token}`,
      },
      body: JSON.stringify(body),
    })

    const result = await response.json() as { code: number }
    return result.code === 0
  }

  /**
   * Create a research progress card.
   */
  static createProgressCard(
    stage: string,
    status: 'running' | 'completed' | 'failed',
    details?: string
  ): FeishuCardMessage {
    const colors: Record<string, FeishuCardColor> = {
      running: 'blue',
      completed: 'green',
      failed: 'red',
    }

    const statusEmoji = {
      running: '🔄',
      completed: '✅',
      failed: '❌',
    }

    return {
      msg_type: 'interactive',
      card: {
        header: {
          title: { tag: 'plain_text', content: `${statusEmoji[status]} Research: ${stage}` },
          template: colors[status],
        },
        elements: [
          ...(details ? [{ tag: 'div' as const, text: { tag: 'lark_md' as const, content: details } }] : []),
          { tag: 'hr' as const },
          { tag: 'note' as const, elements: [{ tag: 'plain_text' as const, content: `Status: ${status} | ${new Date().toISOString()}` }] },
        ],
      },
    }
  }
}
